nav2_rviz_plugins/src/docking_panel.cpp

```diff
@@ -0,0 +1,485 @@
+      if (msg->status_list.back().status == action_msgs::msg::GoalStatus::STATUS_SUCCEEDED) {
+        docking_feedback_indicator_->setText(getDockFeedbackLabel());
+      }
+    });
+
+  undocking_goal_status_sub_ = node->create_subscription<action_msgs::msg::GoalStatusArray>(
+    "undock_robot/_action/status",
+    rclcpp::SystemDefaultsQoS(),
+    [this](const action_msgs::msg::GoalStatusArray::SharedPtr msg) {
+      docking_goal_status_indicator_->setText(
+        nav2_rviz_plugins::getGoalStatusLabel("Feedback", msg->status_list.back().status));
+      if (msg->status_list.back().status != action_msgs::msg::GoalStatus::STATUS_EXECUTING) {
+        docking_button_->setEnabled(true);
+        undocking_button_->setText("Undock robot");
+        undocking_in_progress_ = false;
+      }
+    });
+}
+
+DockingPanel::~DockingPanel()
+{
+}
+
+void DockingPanel::onDockingButtonPressed()
+{
+  if (!docking_in_progress_) {
+    startDocking();
+  } else {
+    cancelDocking();
+  }
+}
+
+void DockingPanel::onUndockingButtonPressed()
+{
+  if (!undocking_in_progress_) {
+    startUndocking();
+  } else {
+    cancelUndocking();
+  }
+}
+
+void DockingPanel::startDocking()
+{
+  auto is_action_server_ready =
+    dock_client_->wait_for_action_server(std::chrono::seconds(5));
+  if (!is_action_server_ready) {
+    RCLCPP_ERROR(client_node_->get_logger(), "dock_robot action server is not available.");
+    return;
+  }
+
+  QComboBox * combo_box = dock_type_;
+  // If "default" option is selected, it gets removed and the next item is selected
+  if (combo_box->findText("Default") != -1) {
+    combo_box->removeItem(0);
+  }
+
+  // If there are no plugins available, return
+  if (combo_box->count() == 0) {
+    return;
+  }
+
+  // Send the goal to the action server
+  auto goal_msg = Dock::Goal();
+  goal_msg.use_dock_id = use_dock_id_;
+  goal_msg.navigate_to_staging_pose = nav_to_staging_checkbox_->isChecked();
+  if (use_dock_id_) {
+    if (dock_id_->text().isEmpty()) {
+      RCLCPP_ERROR(client_node_->get_logger(), "Dock id is empty.");
+      return;
+    }
+    goal_msg.dock_id = dock_id_->text().toStdString();
+
+    RCLCPP_INFO(
+      client_node_->get_logger(), "DockRobot will be called using dock id: %s",
+      goal_msg.dock_id.c_str());
+
+  } else {
+    if (dock_pose_x_->text().isEmpty() || dock_pose_y_->text().isEmpty() ||
+      dock_pose_yaw_->text().isEmpty())
+    {
+      RCLCPP_ERROR(client_node_->get_logger(), "Dock pose is empty.");
+      return;
+    }
+    goal_msg.dock_pose.header.frame_id = "map";
+    goal_msg.dock_pose.header.stamp = client_node_->now();
+    goal_msg.dock_pose.pose.position.x = dock_pose_x_->text().toDouble();
+    goal_msg.dock_pose.pose.position.y = dock_pose_y_->text().toDouble();
+    goal_msg.dock_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+      dock_pose_yaw_->text().toDouble());
+    goal_msg.dock_type = combo_box->currentText().toStdString();
```

Comment:
How is the combo box populated with dock plugins?

Reply:
The combobox is filled using the same ``pluginLoader`` function as the Selector: https://github.com/ros-navigation/navigation2/pull/4458/files#diff-e29dbc8bae4e148962b55e5b26da6a96d5dfca9fdb8223724abca2c463f3dd11R37